base/binaryplatforms.jl

```diff
@@ -7,6 +7,60 @@ export AbstractPlatform, Platform, HostPlatform, platform_dlext, tags, arch, os,
        select_platform, platforms_match, platform_name
 import .Libc.Libdl
 
+### Submodule with information about CPU features
+module CPUID
+
+export cpu_isa
+
+"""
+    ISA(features::Set{String})
+
+A structure which represents the Instruction Set Architecture (ISA) of a
+computer.  It holds the `Set` of features of the CPU.
+"""
+struct ISA
+    features::Set{UInt32}
+end
+
+Base.:<=(a::ISA, b::ISA) = a.features <= b.features
+Base.:<(a::ISA,  b::ISA) = a.features <  b.features
+
+include("features_h.jl")
+
+# Sets of features for the x86-64 microarchitectures come from
+# https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html
+const X86_64_GENERIC = ISA(Set{UInt32}())
+const CORE2 = ISA(Set((X86_64_GENERIC.features..., JL_X86_sse3, JL_X86_ssse3)))
+const NEHALEM = ISA(Set((CORE2.features..., JL_X86_sse41, JL_X86_sse42, JL_X86_popcnt)))
+const SANDYBRIDGE = ISA(Set((NEHALEM.features..., JL_X86_avx, JL_X86_aes, JL_X86_pclmul)))
+const HASWELL = ISA(Set((SANDYBRIDGE.features..., JL_X86_movbe, JL_X86_avx2, JL_X86_fsgsbase, JL_X86_rdrnd, JL_X86_fma, JL_X86_bmi, JL_X86_bmi2, JL_X86_f16c)))
+const SKYLAKE = ISA(Set((HASWELL.features..., JL_X86_clflushopt, JL_X86_xsavec, JL_X86_xsaves)))
+const SKYLAKE_AVX512 = ISA(Set((HASWELL.features..., JL_X86_pku, JL_X86_rdseed, JL_X86_adx, JL_X86_prfchw, JL_X86_avx512f, JL_X86_clwb, JL_X86_avx512vl, JL_X86_avx512bw, JL_X86_avx512dq, JL_X86_avx512cd)))
+const X86_64_ISAS = [SKYLAKE_AVX512, SKYLAKE, HASWELL, SANDYBRIDGE, NEHALEM, CORE2, X86_64_GENERIC]
```

Comment:
I think it would be nice to rephrase this as a big `Dict`, so that it can be easily explored by code.  E.g. something like:

```
ISAs_by_family = Dict(
    "x86_64" => Dict(
         "CORE2" => ISA( ... ),
         ...,
     ),
     "aarch64" => Dict(
         "armv8.1" => ISA( ... ),
         ...,
     ),
     ...
)
```

Then, we can have a `cpu_family()` method that returns something like `"x86_64"`, and `cpu_isa()` becomes just:

```
cpu_isa() = ISA(Set(last(get(ISAs_by_family, cpu_family(), [Set{UInt32}()])))
```